### PR TITLE
Reduce content json size

### PR DESCRIPTION
**Pull request by aurislt7**
imodel02 PR: https://dev.azure.com/bentleycs/iModelTechnologies/_git/imodel02/pullrequest/173021I%2FECPresentation%2FECPresentationTypes.h

### CODE REVIEW

File: presentation/common/src/presentation-common/content/Fields.ts [hunk 1]
```diff
@@ -187,6 +217,47 @@ export class Field {
...
+      relatedClassPath: compressedPropertyJSON.relatedClassPath.map((compressedInfoJSON) => SelectClassInfo.decompressRelatedClassInfoJSON(compressedInfoJSON, classesMap)),
+    };
+  }
+
+  private static decompressPropertyInfoJSON(compressedPropertyJSON: CompressedPropertyInfoJSON, classesMap: { [id: string]: CompressedClassInfoJSON }): PropertyInfoJSON {
+    return {
+      ...compressedPropertyJSON,
+      classInfo: {id: compressedPropertyJSON.classInfo, ...classesMap[compressedPropertyJSON.classInfo]},
+    };
+  }
```
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Fields.ts:259]: I'd move these from class scope to module scope as plain functions.

File: presentation/common/src/presentation-common/content/Descriptor.ts [hunk 2]
```diff
@@ -55,6 +65,26 @@ export namespace SelectClassInfo {
...
+      ...compressedSelectClass,
+      selectClassInfo: {id: compressedSelectClass.selectClassInfo, ...classesMap[compressedSelectClass.selectClassInfo]},
+      navigationPropertyClasses: compressedSelectClass.navigationPropertyClasses.map((compressedInfoJSON) => decompressRelatedClassInfoJSON(compressedInfoJSON, classesMap)),
+      relatedInstanceClasses: compressedSelectClass.relatedInstanceClasses.map((compressedInfoJSON) => decompressRelatedClassInfoJSON(compressedInfoJSON, classesMap)),
+      pathToPrimaryClass: compressedSelectClass.pathToPrimaryClass.map((compressedInfoJSON) => decompressRelatedClassInfoJSON(compressedInfoJSON, classesMap)),
+      relatedPropertyPaths: compressedSelectClass.relatedPropertyPaths.map((path) => path.map((compressedInfoJSON) => decompressRelatedClassInfoJSON(compressedInfoJSON, classesMap))),
+    };
+  }
+
+  export function decompressRelatedClassInfoJSON(compressedInfoJSON: CompressedRelatedClassInfoJSON, classesMap: { [id: string]: CompressedClassInfoJSON }): RelatedClassInfoJSON {
```
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Descriptor.ts:80]: Why is this under `SelectClassInfo` namespace?

File: presentation/common/src/presentation-common/content/Descriptor.ts [hunk 3]
```diff
@@ -55,6 +65,26 @@ export namespace SelectClassInfo {
       relatedInstanceClasses: json.relatedInstanceClasses.map((p) => RelatedClassInfo.fromJSON(p)),
     };
   }
+
+  export function decompressSelectClassInfoJSON(compressedSelectClass: CompressedSelectClassInfoJSON, classesMap: { [id: string]: CompressedClassInfoJSON }): SelectClassInfoJSON {
```
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Descriptor.ts:69]: I see decompression methods defined in two ways:
1. This one is defined under a namespace, named under the thing it decompresses.
2. Some others are defined as private `Descriptor` methods.

We need to choose one or the other, but not both. I'm leaning towards the first approach and naming the methods `fromCompressedJSON`.

File: presentation/common/src/presentation-common/content/Descriptor.ts [hunk 4]
```diff
@@ -130,6 +160,29 @@ export interface DescriptorJSON {
   filterExpression?: string;
 }
 
+/**
+ * Serialized [[Descriptor]] JSON representation.
+ * @beta
+ */
+export interface CompressedDescriptorJSON {
```
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Descriptor.ts:167]: IMO it would be better to omit / add attributes from / to `DescriptorJSON` instead of re-defining everything.

File: presentation/common/src/presentation-common/content/Fields.ts [hunk 5]
```diff
@@ -54,19 +63,40 @@ export interface NestedContentFieldJSON extends BaseFieldJSON {
...
 
+/**
+ * JSON representation of a [[CompressedField]]
+ * @beta
+ */
+export type CompressedFieldJSON = BaseFieldJSON | CompressedPropertiesFieldJSON | CompressedNestedContentFieldJSON;
+
 /** Is supplied field a properties field. */
-const isPropertiesField = (field: FieldJSON | Field): field is PropertiesFieldJSON | PropertiesField => {
+export const isPropertiesField = (field: FieldJSON | Field): field is CompressedPropertiesFieldJSON | PropertiesFieldJSON | PropertiesField => {
```
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Fields.ts:94]: why `export`?

File: presentation/common/src/test/content/Descriptor.test.ts [hunk 6]
```diff
@@ -86,6 +87,23 @@ describe("Descriptor", () => {
       expect(descriptor).to.matchSnapshot();
     });
 
+    it("creates valid descriptorJSON from CompressedDescriptorJSON with nestedField", () => {
+      const compressedDescriptorJSON = compressDescriptorJSON(testDescriptorJSON);
```
Comment (grigasp) [inline presentation/common/src/test/content/Descriptor.test.ts:91]: This test is testing not only the stuff it needs to (`fromJSON` overload with `CompressedDescriptorJSON` argument), but also `compressDescriptorJSON` function, which we don't want.

Instead, I recommend creating a `CompressedDescriptorJSON` object, deserializing `Descriptor` from it and making sure it's okay. The JSON can contain multiple different fields to avoid having two similar tests.

**Review by grigasp — CHANGES_REQUESTED** [review]

File: presentation/common/src/presentation-common/EC.ts [hunk 7]
```diff
@@ -188,6 +197,18 @@ export interface PropertyInfoJSON {
   kindOfQuantity?: KindOfQuantityInfo;
 }
 
+/**
+ * A serialized and compressed version of [[PropertyInfo]]
+ * @public
+ */
+export interface CompressedPropertyInfoJSON {
```
[the remark this answers is not in the document]
Comment (aurislt7) [reply presentation/common/src/presentation-common/EC.ts]: Looks better, changed almost all newly created interfaces using this approach.

File: presentation/common/src/presentation-common/content/Fields.ts (hunk 1 above)
Comment (aurislt7) [reply presentation/common/src/presentation-common/content/Fields.ts]: Moved it.

File: presentation/common/src/presentation-common/content/Descriptor.ts (hunk 2 above)
Comment (aurislt7) [reply presentation/common/src/presentation-common/content/Descriptor.ts]: Moved it to be under RelatedClassInfo namespace

File: presentation/common/src/presentation-common/content/Descriptor.ts (hunk 3 above)
Comment (aurislt7) [reply presentation/common/src/presentation-common/content/Descriptor.ts]: Renamed all the methods to `fromCompressed...`, also moved Descriptor method `fromCompressedJSON` to be under Descriptor namespace.

File: presentation/common/src/presentation-common/content/Descriptor.ts (hunk 4 above)
Comment (aurislt7) [reply presentation/common/src/presentation-common/content/Descriptor.ts]: Updated it

File: presentation/common/src/test/content/Descriptor.test.ts (hunk 6 above)
Comment (aurislt7) [reply presentation/common/src/test/content/Descriptor.test.ts]: Added new tests for `CompressedDescriptorJSON` and `compressDescriptorJSON`. Used `matchSnapshot` for both of them, was that the right approach?

File: presentation/common/src/presentation-common/content/Descriptor.ts [hunk 8]
```diff
@@ -432,3 +388,22 @@ export class Descriptor implements DescriptorSource {
     return overrides;
   }
 }
+
+/** @alpha */
+export namespace Descriptor {
```
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Descriptor.ts:393]: `Descriptor` is a class, so there's no need to add the function to a namespace - you can just add it to `Descriptor` class a a `static` function

File: presentation/common/src/presentation-common/content/Fields.ts [hunk 9]
```diff
@@ -36,67 +35,38 @@ export interface BaseFieldJSON {
...
-/**
- * JSON representation of a [[CompressedField]]
- * @beta
- */
-export type CompressedFieldJSON = BaseFieldJSON | CompressedPropertiesFieldJSON | CompressedNestedContentFieldJSON;
+export type FieldJSON<TClassInfoJSON = ClassInfoJSON> = BaseFieldJSON | PropertiesFieldJSON<TClassInfoJSON> | NestedContentFieldJSON<TClassInfoJSON>;
 
 /** Is supplied field a properties field. */
-export const isPropertiesField = (field: FieldJSON | Field): field is CompressedPropertiesFieldJSON | PropertiesFieldJSON | PropertiesField => {
+export const isPropertiesField = (field: FieldJSON | Field): field is PropertiesFieldJSON | PropertiesFieldJSON<string> | PropertiesField => {
```
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Fields.ts:64]:
```suggestion
export const isPropertiesField = (field: FieldJSON | Field): field is PropertiesFieldJSON<any> | PropertiesField => {
```

File: presentation/common/src/presentation-common/content/Fields.ts [hunk 10]
```diff
@@ -36,67 +35,38 @@ export interface BaseFieldJSON {
...
   return !!(field as any).properties;
 };
 
 /** Is supplied field a nested content field. */
-export const isNestedContentField = (field: FieldJSON | Field): field is CompressedNestedContentFieldJSON | NestedContentFieldJSON | NestedContentField => {
+export const isNestedContentField = (field: FieldJSON | Field): field is NestedContentFieldJSON | NestedContentFieldJSON<string> | NestedContentField => {
```
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Fields.ts:69]:
```suggestion
export const isNestedContentField = (field: FieldJSON | Field): field is NestedContentFieldJSON<any> | NestedContentField => {
```

File: presentation/common/src/test/content/Descriptor.test.ts [hunk 11]
```diff
@@ -87,21 +88,77 @@ describe("Descriptor", () => {
       expect(descriptor).to.matchSnapshot();
     });
 
-    it("creates valid descriptorJSON from CompressedDescriptorJSON with nestedField", () => {
-      const compressedDescriptorJSON = compressDescriptorJSON(testDescriptorJSON);
-      const descriptorFromCompressedJSON = Descriptor.fromJSON(JSON.stringify(compressedDescriptorJSON));
-      const descriptor = Descriptor.fromJSON(JSON.stringify(testDescriptorJSON));
-      expect(descriptor).to.deep.eq(descriptorFromCompressedJSON);
+    it("creates valid Descriptor from CompressedDescriptorJSON", () => {
+      const categories = [createRandomCategoryJSON(), createRandomCategoryJSON(), createRandomCategoryJSON(), createRandomCategoryJSON()];
```
Comment (grigasp) [inline presentation/common/src/test/content/Descriptor.test.ts:92]: Why not just use one category for everything? We're not interested in categories' handling, so there should be as little code as possible related to them.

File: presentation/common/src/test/content/Descriptor.test.snap [hunk 12]
```diff
@@ -1,5 +1,774 @@
...
+          "property": Object {
+            "classInfo": "0x126b50000005c18",
+            "name": "Refined",
+            "type": "string",
+          },
+          "relatedClassPath": Array [
+            Object {
+              "isForwardRelationship": true,
+              "isPolymorphicRelationship": undefined,
+              "isPolymorphicTargetClass": undefined,
```
Comment (grigasp) [inline presentation/common/src/test/content/Descriptor.test.snap:309]: Why do we get `undefined` here? They're not optional in `RelatedClassInfo` interface..

File: presentation/common/src/test/content/Descriptor.test.snap [hunk 13]
```diff
@@ -1,5 +1,774 @@
...
+          "property": Object {
+            "classInfo": Object {
+              "id": "0x1",
+              "label": "label1",
+              "name": "name1",
+            },
+            "name": "mollitia molestiae itaque",
+            "type": "sed doloribus dolorem",
+          },
+          "relatedClassPath": Array [],
```
Comment (grigasp) [inline presentation/common/src/test/content/Descriptor.test.snap:567]: The test sets this value, but it's empty here

File: presentation/common/src/test/content/Descriptor.test.ts [hunk 14]
```diff
@@ -87,21 +88,77 @@ describe("Descriptor", () => {
...
+      const ids = ["0x1", "0x2", "0x3", "0x4"];
+      const testRelatedClassInfo: RelatedClassInfoJSON<string> = {
+        sourceClassInfo: ids[1],
+        targetClassInfo: ids[2],
+        relationshipInfo: ids[3],
+        isForwardRelationship: true,
+      };
+      const fields: FieldJSON<string>[] = [
+        createRandomPrimitiveFieldJSON(categories[0]), {
```
Comment (grigasp) [inline presentation/common/src/test/content/Descriptor.test.ts:101]: What's the point of adding a field that doesn't require compressing / decompressing?

File: presentation/common/src/test/content/Descriptor.test.snap [hunk 15]
```diff
@@ -1476,3 +2245,194 @@ Object {
   "sortingField": undefined,
 }
 `;
+
+exports[`Descriptor fromJSON creates valid descriptorJSON from CompressedDescriptorJSON 1`] = `
```
Comment (grigasp) [inline presentation/common/src/test/content/Descriptor.test.snap:2249]: I don't see a test for this

File: presentation/common/src/presentation-common/EC.ts [hunk 16]
```diff
@@ -176,12 +185,8 @@ export namespace PropertyInfo {
   }
 }
 
-/**
- * A serialized version of [[PropertyInfo]]
- * @public
- */
-export interface PropertyInfoJSON {
-  classInfo: ClassInfoJSON;
+export interface PropertyInfoJSON<TClassInfoJSON = ClassInfoJSON> {
```
Comment (grigasp) [inline presentation/common/src/presentation-common/EC.ts:188]: Class header is gone

File: presentation/common/src/presentation-common/EC.ts [hunk 17]
```diff
@@ -236,6 +241,16 @@ export namespace RelatedClassInfo {
     };
   }
 
+  /** Deserialize compressed [[RelatedClassInfo]] from JSON */
```
Comment (grigasp) [inline presentation/common/src/presentation-common/EC.ts:244]: I'd reword this to `Deserialize [[RelatedClassInfo]] from compressed JSON`

File: presentation/common/src/presentation-common/content/Descriptor.ts [hunk 18]
```diff
@@ -55,6 +55,17 @@ export namespace SelectClassInfo {
       relatedInstanceClasses: json.relatedInstanceClasses.map((p) => RelatedClassInfo.fromJSON(p)),
     };
   }
+
+  export function fromCompressedSelectClassInfoJSON(compressedSelectClass: SelectClassInfoJSON<string>, classesMap: { [id: string]: CompressedClassInfoJSON }): SelectClassInfoJSON {
```
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Descriptor.ts:59]: The function is already under `SelectClassInfo` namespace - no need to repeat that in function name

File: presentation/common/src/presentation-common/EC.ts [hunk 19]
```diff
@@ -236,6 +241,16 @@ export namespace RelatedClassInfo {
...
+  export function fromCompressedRelatedClassInfoJSON(compressedInfoJSON: RelatedClassInfoJSON<string>, classesMap: { [id: string]: CompressedClassInfoJSON }): RelatedClassInfoJSON {
```
Comment (grigasp) [inline presentation/common/src/presentation-common/EC.ts:245]: The function is already under `RelatedClassInfo` namespace - no need to repeat that in function name

File: presentation/common/src/presentation-common/content/Descriptor.ts [hunk 20]
```diff
@@ -130,6 +141,19 @@ export interface DescriptorJSON {
   filterExpression?: string;
 }
 
+/**
+ * Serialized [[Descriptor]] JSON representation.
+ * @beta
```
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Descriptor.ts:146]: Can't be `@beta` if used in `@public` functions

File: presentation/common/src/presentation-common/content/Fields.ts [hunk 21]
```diff
@@ -187,6 +187,16 @@ export class Field {
     });
   }
 
+  public static fromCompressedFieldJSON(json: FieldJSON<string>, classesMap: { [id: string]: CompressedClassInfoJSON }): FieldJSON | undefined {
```
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Fields.ts:190]: The function is already under `Field` class - no need to repeat that in function name. Also, missing comment.

**Review by grigasp — CHANGES_REQUESTED** [review]

File: presentation/common/src/presentation-common/content/Descriptor.ts [hunk 22]
```diff
@@ -55,6 +55,17 @@ export namespace SelectClassInfo {
       relatedInstanceClasses: json.relatedInstanceClasses.map((p) => RelatedClassInfo.fromJSON(p)),
     };
   }
+
+  export function fromCompressedJSON(compressedSelectClass: SelectClassInfoJSON<string>, classesMap: { [id: string]: CompressedClassInfoJSON }): SelectClassInfoJSON {
```
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Descriptor.ts:59]: public functions require docs

**Review by grigasp — CHANGES_REQUESTED** [review]

File: presentation/common/src/presentation-common/content/Fields.ts (hunk 5 above)
Comment (aurislt7) [reply presentation/common/src/presentation-common/content/Fields.ts]: It's no more needed, fixed it

File: presentation/common/src/presentation-common/EC.ts [hunk 23]
```diff
@@ -179,6 +179,17 @@ export namespace PropertyInfo {
     return { ...info, classInfo: ClassInfo.toJSON(info.classInfo) };
   }
 
+  /** Serialize [[PropertyInfoJSON]] to compressed JSON */
+  export function toCompressedJSON(json: PropertyInfoJSON, classesMap: { [id: string]: CompressedClassInfoJSON }): PropertyInfoJSON<string> {
```
Comment (grigasp) [inline presentation/common/src/presentation-common/EC.ts:183]: shouldn't this take `PropertyInfo` instead of `PropertyInfoJSON`?

Comment (aurislt7) [reply presentation/common/src/presentation-common/EC.ts]: Yep, it should. fixed it

**Review by grigasp — APPROVED** [review]

**Review by grigasp — APPROVED** [review]
